File: game/forest/hurt_rock_talking.py
import pygame

from animation import Animation
from audio_helper import AudioHelper
from forest.forest_resources import ForestResources
from forest.forest_utils import ForestUtils
from game_data import GameData
from render_type import RenderType
from state import State


class HurtRockTalking(State):
    def __init__(self, context: GameData):
        super().__init__(context)
        self.needs_bottom = RenderType.POST

    def process_events(self, events):
        if self.get_frame_index() >= len(ForestResources.sync_rock):
            return ForestUtils.reduce_lives(self.context)

        return None

    def on_enter(self) -> None:
        super().on_enter()
        AudioHelper.play(ForestResources.speak_rock, self.context.country)

    def render(self, screen):
        screen.blit(Animation.get_sync_frame(ForestResources.hit_rock_sync, ForestResources.sync_rock, self.get_frame_index()), (0,0))

File: game-web/src/forest/hurtRockTalking.ts
/** Port of game/forest/hurt_rock_talking.py */

import { State, type StateClass } from '../state';
import { Animation } from '../animation';
import { AudioHelper } from '../audioHelper';
import { ForestResources } from './forestResources';
import { ForestUtils } from './forestUtils';
import { RenderType } from '../renderType';
import type { GameData } from '../gameData';
import type { PhoneEvents } from '../phoneEvents';

export class HurtRockTalking extends State {
  readonly needsBottom = RenderType.POST;

  constructor(context: GameData) { super(context); }

  processEvents(_events: PhoneEvents): StateClass | null {
    if (this.getFrameIndex() >= ForestResources.syncRock.length) {
      return ForestUtils.reduceLives(this.context, 3); // Rock
    }
    return null;
  }

  onEnter(): void {
    super.onEnter();
    AudioHelper.play(ForestResources.speakRock);
  }

  render(ctx: OffscreenCanvasRenderingContext2D): void {
    ctx.drawImage(Animation.getSyncFrame(ForestResources.hitRockSync, ForestResources.syncRock, this.getFrameIndex()), 0, 0);
  }
}
